src/Workspaces/Core/Portable/Telemetry/TelemetryLogging.cs

```diff
@@ -3,26 +3,49 @@
 // See the LICENSE file in the project root for more information.
 
 using System;
+using System.Threading;
+using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Telemetry;
 
 /// <summary>
 /// Provides access to posting telemetry events or adding information
-/// to aggregated telemetry events.
+/// to aggregated telemetry events. Posts pending telemetry at 30
+/// minute intervals.
 /// </summary>
 internal static class TelemetryLogging
 {
     private static ITelemetryLogProvider? s_logProvider;
+    private static AsyncBatchingWorkQueue? s_postTelemetryQueue;
 
     public const string KeyName = "Name";
     public const string KeyValue = "Value";
     public const string KeyLanguageName = "LanguageName";
     public const string KeyMetricName = "MetricName";
 
-    public static void SetLogProvider(ITelemetryLogProvider logProvider)
+    public static event EventHandler<EventArgs>? Flushed;
+
+    public static void SetLogProvider(ITelemetryLogProvider logProvider, IAsynchronousOperationListener asyncListener)
     {
         s_logProvider = logProvider;
+
+        if (s_postTelemetryQueue is null)
+        {
+            var postTelemetryQueue = new AsyncBatchingWorkQueue(
+                TimeSpan.FromMinutes(30),
+                PostCollectedTelemetryAsync,
+                asyncListener,
+                CancellationToken.None);
+
+            if (null == Interlocked.CompareExchange(ref s_postTelemetryQueue, postTelemetryQueue, null))
+            {
+                // We created the work queue in use. Add an item into it to kick things off.
+                s_postTelemetryQueue?.AddWork();
```

Comment:
💡 It's fine if this gets called more than once. The work queue will aggregate duplicate requests into a single one.